### ISSUE TITLE
view::zip with view::for_each fails to compile (another case)

### ISSUE THREAD

**Issue by etam** [ericniebler/range-v3#441]
Here's the minimal example:

```
#include <vector>

#include <range/v3/core.hpp>
#include <range/v3/view/for_each.hpp>
#include <range/v3/view/zip.hpp>

int main()
{
    using namespace ranges;

    const auto v1 = std::vector<int>{1, 2, 3};
    const auto v2 = std::vector<char>{'a', 'b', 'c'};

    view::zip(v1, v2) | view::for_each([](const auto& value) { return yield(value); });
}
```

output from gcc 6.1.1: http://susepaste.org/533f6b44
output from clang 3.8.0: http://susepaste.org/c7b2e98f

**Comment by ericniebler** [comment]
Looking at CaseyCarter/range-v3@3b6544cb7b6d64446e4706d7ac433081a23177ed, is the problem that the `pair` created by the `zip` view not copy constructible?

**Comment by CaseyCarter** [comment]
Yes, the basic problem is that `view::single` requires a `SemiRegular` type, and the reference type of a zip view is (generally) not `SemiRegular` because it has reference members. This is sort-of fixed by pushing the error up to `yield`.

On a higher level, the root cause is that proxy references don't decay to their value types as do core language references when passed to a function template that accepts parameters by value. Fixing this - indeed deciding if it needs to *be* fixed - is research.

**Comment by aldanor** [comment]
What would the minimal suggested change be for the example above then?

**Comment by FloopCZ** [comment]
Hi, the minimal possible change might be to explicitly specify the lambda argument type to something copy constructible:
```
view::zip(v1, v2) | view::for_each([](std::pair<int, char> value) { return yield(value); });
```
@ericniebler Maybe zip could return std::reference_wrapper instead of int& or char&?

**Comment by FloopCZ** [comment]
So with PR #559, one can yield a pair of references:
```
view::zip(v1, v2) 
| view::for_each([](std::pair<const int&, const char&> value) { return yield(value); });
```
However, it is not possible to use automatic type deduction for the lambda argument, as requested by the issuer:
```
view::zip(v1, v2)
| view::for_each([](const auto& value) { return yield(value); });
```
Any ideas why the lambda argument type cannot be deduced automatically?

**Comment by FloopCZ** [comment]
ok, I dug into the issue and the problem seems to be with the IndirectInvocable concept. If I comment out the concept check in view/for_each.hpp:63
```
IndirectInvocable<F, range_iterator_t<Rng>>
```
everything seems to compile and work as expected.

To be more specific, it is sufficient to uncomment the second part of the IndirectInvocable concept in utility/iterator_concepts.hpp:498-502, which says:
```
            meta::lazy::invoke<
                detail::iter_map_reduce_fn_<
                    meta::bind_front<meta::quote<concepts::Invocable::result_t>, C&>,
                    meta::quote<CommonReference>>,
                Is...>
```

**Comment by ericniebler** [comment]
This suggests a bug in `CommonReference`. Perhaps ambiguous partial specializations of `basic_common_reference`. I can't check right now, though.

**Comment by FloopCZ** [comment]
Should we really close this before the CommonReference problem is fixed?

**Comment by CaseyCarter** [comment]
> This suggests a bug in `CommonReference`.

IIRC, the type of `yield(value)` is `repeat_n_view<decay_t<decltype(value)>>`. Although the different argument types passed to the generic lambda:
```c++
[](const auto& value) { return yield(value); }
```
may share a common reference type, there's no reason to believe that the *return types* will do so. This isn't a bug in `CommonReference`.

I think the root issue here is impedance mismatch resulting from trying to create a range of values whose value type is the reference type of some other range. `pair<T, U>` and `tuple<Ts...>` cause all sorts of weirdness in that they are sometimes value types, and sometimes reference types, and neither `std::decay_t` nor template argument deduction know what's going on.

If we had a customization point that knew how to turn reference types into value types - perhaps wrapped in a `view::decay` adapter - the OP could use `view::zip(v1, v2) | view::decay | view::for_each([](const auto& value) { return yield(value); });`.

**Comment by FloopCZ** [comment]
Just to be clear, does it mean that tuple<int&, int&> would decay to tuple<int, int>? And is it really what we want? Because if one of the zipped ranges contains e.g., std::future or std::unique_ptr, one may really need the reference type.

**Comment by CaseyCarter** [comment]
> Just to be clear, does it mean that `tuple<int&, int&>` would decay to `tuple<int, int>`?
Yes.

> And is it really what we want?

That would be the point of `view::decay`, yes ;)

> Because if one of the zipped ranges contains e.g., `std::future` or `std::unique_ptr`, one may really need the reference type.

Yes, but capturing references from iterators is incredibly dangerous: the Ranges TS design doesn't in general guarantee that a reference obtained from an iterator is valid outside the lifetime of that iterator, or after modifying the value of the specific iterator object from which the reference was obtained. It shouldn't be *impossible* to store references, but it should require explicit action on the user's part and not be something that happens implicitly.

**Comment by FloopCZ** [comment]
That is a valid point. :-)
However, const references (as asked by the OP) should be legal, right?

**Comment by CaseyCarter** [comment]
> However, const references (as asked by the OP) should be legal, right?

The issue isn't forming references like the `const auto&` parameter to the lambda in the OP, it's storing references and accessing them later. The lambda in the OP wants to store the reference into a `single_view` which - since it has value semantics - is expected to remain valid "forever."

**Comment by FloopCZ** [comment]
Oh, I can see the problem with single_view, thank you Casey.

In my opinion, it may be confusing and easily forgotten for the users to have to write view::decay before every view::for_each preceded by a view::zip. Are there any alternatives?

Maybe to change the behavior of view::zip, e.g, if zip's parameter is an lvalue container, build a reference, if zip's parameter is a view, build a value. Then one could safely allow the intuitive code proposed by OP.

**Comment by CaseyCarter** [comment]
> it may be confusing and easily forgotten for the users to have to write view::decay

Yes, having people opt-out of dangerous behavior is a recipe for disaster. I think a model with positive identification of capital-R References (both core language reference types and proxy reference types) may be the better solution. We require types to tell us when they are iterators, it may be necessary for types that are references to declare that fact as well.

**Comment by ericniebler** [comment]
> We require types to tell us when they are iterators, it may be necessary for types that are references to declare that fact as well.

This is becoming increasingly clear.

**Comment by FloopCZ** [comment]
That sounds like the only option.

So for instance: All the types are values by default. The only exception is, when a view is created from an lvalue container (this only makes sense for the leftmost view in a pipeline). In such a case, it wraps the container values using std::ref first. The created std::reference_wrappers are preserved through all the views in the pipeline. In views such as view::transform and view::for_each, the reference wrappers are unpacked to regular references and passed to the provided function, however, the user may easily pack them back in the return expression using std::ref (explicitly!). Views such as remove_if would also naturally unpack the reference_wrappers prior to applying the predicate. Views such as view::zip would work out-of-the-box, since their make_tuple would only convert std::reference_wrapper to a regular reference packed in a tuple and thus the reference semantics would be preserved.

With this functionality, one could safely zip views and containers together and easily create proxy containers of std::reference_wrappers, which could be reversed, filtered and whatnot using views.

I could imagine such a functionality, however, it would probably bring many breaking changes.
